pkg/resource/claim_defaulting_reconciler.go

```diff
@@ -0,0 +1,122 @@
+// set, but which one is undefined.
+type ClaimDefaultingReconciler struct {
+	client    client.Client
+	newClaim  func() Claim
+	classKind ClassKind
+}
+
+// NewClaimDefaultingReconciler returns a ClaimDefaultingReconciler that sets
+// the class reference of a resource claim to the resource class annotated as
+// the default.
+func NewClaimDefaultingReconciler(m manager.Manager, of ClaimKind, to ClassKind) *ClaimDefaultingReconciler {
+	nc := func() Claim { return MustCreateObject(schema.GroupVersionKind(of), m.GetScheme()).(Claim) }
+
+	// Panic early if we've been asked to reconcile a claim or resource kind
+	// that has not been registered with our controller manager's scheme.
+	_ = nc()
+
+	return &ClaimDefaultingReconciler{
+		client:    m.GetClient(),
+		newClaim:  nc,
+		classKind: to,
+	}
+}
+
+// Reconcile a resource claim by using its class selector to select and allocate
+// it a resource class.
+func (r *ClaimDefaultingReconciler) Reconcile(req reconcile.Request) (reconcile.Result, error) {
+	log.V(logging.Debug).Info("Reconciling", "controller", claimDefaultingControllerName, "request", req)
+
+	ctx, cancel := context.WithTimeout(context.Background(), claimDefaultingReconcileTimeout)
+	defer cancel()
+
+	claim := r.newClaim()
+	if err := r.client.Get(ctx, req.NamespacedName, claim); err != nil {
+		// There's no need to requeue if we no longer exist. Otherwise we'll be
+		// requeued implicitly because we return an error.
+		return reconcile.Result{}, errors.Wrap(IgnoreNotFound(err), errGetClaim)
+	}
+
+	// There could be several controllers racing to schedule this claim. If it
+	// was scheduled since we were queued then another controller won and we
+	// should abort.
+	if claim.GetClassReference() != nil {
+		return reconcile.Result{Requeue: false}, nil
+	}
+
+	classes := &unstructured.UnstructuredList{}
+	classes.SetGroupVersionKind(r.classKind.List())
+
+	if err := r.client.List(ctx, classes); err != nil {
+		// Claim defaulting controllers don't update the synced status because
+		// no one scheduler has the full view of whether the process failed or
+		// succeeded. It's possible another controller can successfully set a
+		// class even though we can't, so it would be confusing to mark this
+		// claim as failing to be reconciled. Instead we return an error - we'll
+		// be requeued but abort immediately if the claim was defaulted.
+		return reconcile.Result{}, errors.Wrap(err, errListClasses)
+	}
+
+	defaults := []unstructured.Unstructured{}
+	for _, c := range classes.Items {
+		if c.GetAnnotations()[v1alpha1.AnnotationDefaultClassKey] == v1alpha1.AnnotationDefaultClassValue {
+			defaults = append(defaults, c)
+		}
+	}
+
+	if len(defaults) == 0 {
+		// None of our classes are annotated as the default.
+		// There's nothing for us to do.
+		return reconcile.Result{Requeue: false}, nil
+	}
+
+	claim.SetClassReference(meta.ReferenceTo(&defaults[0], schema.GroupVersionKind(r.classKind)))
```

Comment:
Can we add that `TODO` you mentioned here? Alternatively, I believe this would be fixed just by mimicking the claim scheduling reconciler randomness:
```go
random := rand.New(rand.NewSource(time.Now().UnixNano()))
selected := defaults[random.Intn(len(defaults))]
claim.SetClassReference(meta.ReferenceTo(&selected, schema.GroupVersionKind(r.classKind)))
```

Reply:
Okay fiiine I just implemented this.